### ISSUE TITLE
Interested in WebSockets support?

### ISSUE THREAD

**Issue by gngeorgiev** [go-kit/kit#132]
Are you interested in supporting websockets, I need to have a webosocket connection between my microservices and I would love to be able to use gokit with all of its tooling. If yes, I can work on it, and submit a PR.

**Comment by peterbourgon** [comment]
Can you be more specific about "support"? Do you mean providing an Endpoint adapter for a Websocket transport?

**Comment by gngeorgiev** [comment]
What I mean is if someone should come to go-kit today he should be able to use websockets just as he would use json over http, I haven't researched in dept but this is my idea. Let me know if this is not clear.

**Comment by peterbourgon** [comment]
Websockets are not the same as JSON over HTTP :) What are you trying to accomplish? RPC? Streaming? Both?

**Comment by gngeorgiev** [comment]
Yes, I am aware that it is not the same, what I mean is that it should have the same grade of support as the above mentioned. What I need is one of my services to be able to push data to other services which will initiate a websocket connection to it. The catch is that this service does not know how to find all other services which are managed by kubernetes, so they are pretty dynamic, thus the websocket need.

**Comment by peterbourgon** [comment]
I'm sorry, I'm still not really following what you want. Can you be more specific about what you'd like Go kit to provide? Simply providing a websocket transport doesn't seem like it will be enough?

**Comment by anarcher** [comment]
I'm interested in using `websocket` with gokit. (Specifically for the bidirectional use case.)
According to my limited experience with [kite](https://github.com/koding/kite), a bidirectional communication approach should work with the RPC message pattern. (The main reason I use websocket is for bidirectional purposes.)
I'm not sure if `Endpoint` type should be changed for this or not, but I can make "WebsocketRequest" and use "WebsocketRequest.Client.Request" and return  the `Client` to my service... (IMO, `Endpoint` type shouldn't changed for this.)

https://github.com/koding/kite/blob/387abd65d515285e87fa968b666d4ed18ac05764/examples/math/math.go#L51
```
func Square(r *kite.Request) (interface{}, error) {
	// Unmarshal method arguments
	a := r.Args.One().MustFloat64()

	result := a * a

	fmt.Printf("Call received, sending result %.0f back\n", result)

	// Print a log on remote Kite.
	// This message will be printed on client's console.
	r.Client.Go("kite.log", fmt.Sprintf("Message from %s: \"You have requested square of %.0f\"", r.LocalKite.Kite().Name, a))

	// You can return anything as result, as long as it is JSON marshalable.
	return result, nil
}
```

**Comment by yookoala** [comment]
@anarcher The example you provided seems to be an RPC server endpoint on websocket. It could be achieved with a proper Websocket transport.

How about the client side code? Do you have example of it?

**Comment by anarcher** [comment]
@yookoala According to `kite`, The client and the server have the same interface. This means that `kite.HandleFunc` is used in both the client and the server. Of course, to receive a request at the client,  `kite.Run` is called at the client side.

**Comment by chaudharisuresh997** [comment]
> Are you interested in supporting websockets, I need to have a webosocket connection between my microservices and I would love to be able to use gokit with all of its tooling. If yes, I can work on it, and submit a PR.

Did u implement Websocket support for gokit where it is now
